### ISSUE TITLE
Out of memory/leak when using StreamReader::iterator

### ISSUE THREAD

**Issue by cjh1** [OpenChemistry/stempy#22]
The following quickly exhaust all available memory. Not sure if its a leak or we are just loading to much into memory. 
```python
files = ['/data/4dstem/smallScanningDiffraction/data0000.dat'] * 100

reader = io.reader(files)
img = image.create_stem_image(reader, 160, 160,  40, 288);

```

**Comment by psavery** [comment]
I first tried this on my laptop (which has an SSD), first with OpenMP and then with CUDA. 

With OpenMP, I actually didn't run out of memory. It just keep going and was able to process the data and finish the run (I tried it around 8 times). It would usually stall around ~ 2GB of memory usage during the run.

With CUDA, I ran out of memory. 

For the HDD on ulex, I was also able to run OpenMP fine, but CUDA ran out of memory.

My theory: what appears to be happening is that CUDA (at least on my computer and ulex) can't process the data here as fast as it is being read, but OpenMP can. In our processing time benchmarks, OpenMP was nearly 8 times faster than CUDA.

Now, why can't CUDA keep up? I think the answer is disk caching. I think it reads in the data very quickly after the first read, and OpenMP can keep up with the fast reading from the disk cache, but CUDA can't.

I actually also just tried clearing the disk cache [here](https://github.com/OpenChemistry/stempy/blob/35b6c19631fbee7f94287dd275f0d1ad2183e605/stempy/reader.cpp#L47) to confirm this theory, and indeed, the CUDA run did not run out of memory. So I think disk caching is the reason for the memory overload, and CUDA not being able to keep up.

This problem does bring up an important question, though: should we do something to make sure we don't read in too much memory at once? Maybe the main thread can wait for some of the data to process first? It probably doesn't matter in cases when we are I/O bound, but it might matter if we are not.

Also, I guess disk caching might give us a situation where we *aren't* I/O bound for CUDA (at least with our GPUs, more powerful GPUs might be better) and serial. However, I guess OpenMP is so fast that it's even still I/O bound with disk caching...

**Comment by cjh1** [comment]
@psavery Thank you for this detail report. To give you some more background, I experienced this out of memory situation when run on a node on cori, trying to process 164 files. In this situation I would still expect the process to be IO bound so was surprised to see this ( this is with OpenMP ). I then tried to reproduce the problem on my machine and was able to drive the memory usage to a point when it ground to a halt, again surprising if we are IO bound. I will try to reproduce my runs.
